---
title: "KMA712 Assignment 2"
author: "Bindoff, A."
output: html_document
---

`r Sys.time()`

3 week Extension granted (email correspondence 12th June 2019)  

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

```{r}
# Example data
library(microbiome)
library(dplyr)
library(table1)
library(captioner)

data(dietswap)

# Make sure we use functions from correct package
transform <- microbiome::transform

#devtools::install_github("hrbrmstr/hrbrthemes")
library(hrbrthemes)
library(gcookbook)
library(tidyverse)


shuffle <- function(x, set.seed = 42){
  set.seed(set.seed)
  sample(x, length(x), replace = FALSE)
}

#pal <- shuffle(scico::scico(n = 17, palette ='berlin'), set.seed = 99)
pal <- shuffle(c(scico::scico(n = 11, palette ='berlin'),
         scico::scico(n = 11, palette ='hawaii')), set.seed = 99)

```

## Background {.tabset}

Colon cancer is a leading cause of cancer death in North America and Europe. Despite advances in early detection widely available in North America, African Americans (AAM) have a far higher incidence of death by colon cancer (65:100,000) than rural Africans (AFR), at <5:100,000 [@o2015fat]. Although colon cancer risk is determined by many factors, diet may explain some of the differences between AAM and AFR.  

In order to investigate differences in gut microbiota potentially caused by diet between AAM and AFR, O'Keefe and colleagues collected stool samples before and after a dietary intervention in age and sex matched samples of healthy AAM and AFR participants [@o2015fat]. A HITChip phylogenetic microarray was used to profile microbiota composition.  

For this reproducible workflow we obtained data from the O'Keefe study, which was included in the `microbiome` R package [@lahti2019]. Data were prepared loosely following the tutorial [Introduction to the microbiome R package](http://microbiome.github.io/microbiome/Composition.html), filtering on pre-intervention and taking the first sample collected within each group in order to avoid pseudo-replication considerations. We also filtered on the most prevalent taxa to reduce the data in order to aid interpretation and reduce computation time.  

```{r echo = TRUE}
# Use most prevalent taxa to speed up examples
# detection threshold >50 otus, prevalence > .4
pseq <- core(dietswap, detection = 50, prevalence = .4)

# choose sample subset, first sample collection pre-intervention
library(phyloseq)
pseq2 <- subset_samples(pseq, group == "HE" & timepoint.within.group == 1)

# normative western adults for comparison
data(atlas1006)
pseq3 <- atlas1006 %>%
          subset_samples(DNA_extraction_method == "r") %>%
          aggregate_taxa(level = "Phylum") %>%  
          microbiome::transform(transform = "compositional")

```

```{r}
# working with phyloseq objects is not difficult but 
# for some tasks without `phyloseq` or `microbiome` methods
# a data.frame or tibble can be helpful for a novice like me

otu <- tibble::rownames_to_column(as.data.frame(t(otu_table(pseq2)))) %>%
  mutate(sample = rowname) %>%
  select(-rowname)

d <- as_tibble(sample_data(pseq2)) %>% 
  left_join(otu)

label(d$sex) <- "Gender"
label(d$bmi_group) <- "BMI category"

table_nums <- captioner::captioner(prefix = "Table")
tab.1 <- table_nums(name = "tab_1",
                    caption = "Demographics, African American (AAM) and rural African (AFR) participants")
fig_nums <- captioner::captioner(prefix = "Figure")
fig.1 <- fig_nums(name = "fig_1",
                  caption = "Abundance of each taxa, sorted by bacteroidetes")

fig.2 <- fig_nums(name = "fig_2",
                  caption = "Abundance of genera in each sample, sorted by study groups")

fig.3 <- fig_nums(name = "fig_3",
                  caption = "Average abundance of genera for each BMI group")

```

`r table_nums('tab_1')`

```{r table_1, fig.cap = tab.1}

table1(~ sex +  bmi_group | nationality, d)

```
  
  
`r fig_nums('fig_1')`


```{r fig_1, fig.cap = fig.1}
theme_set(theme_bw(21))
p <- pseq3 %>%
    plot_composition(sample.sort = "Bacteroidetes", otu.sort = "abundance", alpha = 0.3) +
         # Set custom colors
          scale_fill_manual(values = default_colors("Phylum")[taxa(pseq3)]) +
  facet_wrap(~Tax, scales = "free_y") +
  theme(text = element_text(size = 8),
    axis.text.x = element_blank())


print(p)

```

Relative abundance of core taxa sorted by study group (AAM and AFR) shows differential abundance of genera. Note relative abundance of *Prevotella melaninogenica* (shaded in pink), which is more abundant in the AFR group.  

`r fig_nums('fig_2')`

```{r genera, fig.cap = fig.2}

# Limit the analysis on core taxa and specific sample group
p <- plot_composition(pseq2,
              taxonomic.level = "Genus",
                      sample.sort = "nationality",
                      x.label = "nationality") +
     guides(fill = guide_legend(ncol = 1)) +
   scale_fill_manual(values = pal) +
     scale_y_percent() +
     labs(x = "Samples", y = "Relative abundance (%)",
                                   title = "Relative abundance data",
                                   subtitle = "",
                                   caption = "") + 
    # theme_ipsum(grid="Y") +
  theme(text = element_text(size = 8),
    axis.text.x = element_text(size = 6, angle = 90),
        legend.key.height = unit(2, "mm"))
print(p)  
```

Averaging abundance by BMI group shows a greater abundance of *Prevotella melaninogenica* in lean participants, but Table 1 shows more lean rural Africans than African Americans so this pattern must be interpreted with caution.  

`r fig_nums('fig_3')`  

```{r bmi_group, fig.cap = fig.3}

# Averaged by group
p <- plot_composition(pseq2,
                      average_by = "bmi_group", transform = "compositional") +
  scale_fill_manual(values = pal) +
  theme(text = element_text(size = 10),
        legend.key.height = unit(4, "mm"))
print(p)
```

*Ordination* reduces high-dimension data to lower-dimensional data, such that objects that are similar to each other are projected onto a lower-dimensional space 'nearer' to each other and objects that are dissimilar are projected further apart. In `r fig_nums('fig_3')`, *samples* are sorted along a single dimension using non-metric multi-dimensional scaling (NMDS) (on Bray-Curtis distances; see `help(neatsort)` for available methods in the `microbiome` R package), then plotted as a heatmap of abundances. This shows some distinct clusters of *Prevotella melaninogenica*, *Bacteroides vulgatus*, and *Oscillospira guillermondii*. However, no comparison between groups can be easily made from this.  

```{r}
p <- plot_composition(microbiome::transform(pseq2, "compositional"),
                    plot.type = "heatmap",
                    otu_sort = "neatmap") +
  scico::scale_fill_scico(palette = "bilbao", name = "") +
  theme(text = element_text(size = 8),
        axis.text.x = element_text(size = 6, angle = 45, hjust = 1),
        axis.text.y = element_text(size = 5))
print(p)
```

Continuing with the ordination theme, we now compute a distance matrix of genera based on sample similarities then reduce these distances to a two-dimensional space using t-distributed stochastic neighbour embeddings (tSNE). Taxa that appear in similar compositions in samples should be nearer to each other than taxa that are dissimilar.  

```{r}
library(Rtsne)
library(vegan)

# distance matrix (sample similarities)
ps <- microbiome::transform(pseq2, "hellinger")
dm <- vegdist(otu_table(ps), "euclidean")

m1 <- Rtsne(dm, dims = 2, perplexity = 7, theta = 0.1, max_iter = 5000, verbose = F)

proj <- m1$Y

rownames(proj) <- rownames(otu_table(ps))

proj <- tibble::rownames_to_column(as.data.frame(proj))

ggplot(proj, aes(x = V1, y = V2, label = rowname)) +
  geom_text(nudge_y = -1.5, size = 2.5) +
  geom_point()
```

Using a similar approach, we ordinate samples on two dimensions using tSNE. This could be considered an *unsupervised statistical learning* method, because tSNE doesn't know about nationality, BMI etc, only abundance of taxa in each sample. In fig______ we plot the ordination using symbols for nationality and shade with a genera of interest.  

```{r}
A <- apply(d[, c(9:25)], 2, scale)  # standardize across taxa

set.seed(42)
m2 <- Rtsne(A, dims = 2, perplexity = 9, theta = 0.05, max_iter = 5000, verbose = F)

B1 <- data.frame(d,
                 m2$Y)
ggplot(B1, aes(x = X1, y = X2,
               shape = nationality,
               colour = `Prevotella.melaninogenica.et.rel.`)) +
  geom_point(size = 3) +
  scico::scale_colour_scico(palette = "vik") +
  theme_minimal()
```

This realisation of the data clearly shows two clusters, one composed entirely of samples from African-Americans with low abundance of *Prevotela melaninogenica* and the other composed mostly of rural Africans with high abundance of *Prevotela melaninogenica*. I say "this realisation of the data", because tSNE has a hyper-parameter called "perplexity" which can produce variation. explain perplexity, describe the next figure, explain that it hasn't made any difference to the similarity, but the components are meaningless



```{r}
set.seed(42)

perplexity_iter <- function(x, p){
  Rtsne(x, dims = 2, perplexity = p, max_iter = 2000, theta = 0.05, verbose = FALSE)
}


m <- lapply(c(4,9,12), function(p) perplexity_iter(x = A, p = p))

extract_ys <- function(m){
  perplexity <- m$perplexity
  Y1 <- m$Y[,1]
  Y2 <- m$Y[,2]
  cbind(d, Y1, Y2, perplexity)
}

k <- bind_rows(lapply(m, extract_ys))

```

```{r}
library(gganimate)
ggplot(k, aes(x = Y1, y = Y2, shape = nationality, colour = `Prevotella melaninogenica et rel.`)) +
  geom_point(size = 4) +
  scico::scale_colour_scico(palette = "vik") +
  theme_minimal() +
  transition_states(perplexity) +
  ease_aes('linear') +
  view_follow() +
  labs(title = 'Perplexity = {closest_state}')

```

```{r}
foo <- function(x){
  d %>% mutate(col = scale(unlist(d[, x])),
               shade = x,
                    Y1 = m2$Y[, 1],
                    Y2 = m2$Y[, 2])
  
}

k <- lapply(names(d[, 9:25]), foo)
k <- dplyr::bind_rows(k)

ggplot(k, aes(x = Y1, y = Y2, shape = nationality, colour = col)) +
  geom_point(size = 5) +
  scico::scale_colour_scico(palette = "vik") +
  theme_minimal() +
  transition_states(shade) +
  ease_aes('sine-in-out') +
  labs(title = 'Taxa shading = {closest_state}')

```



```{r}

D <- dist(A)
m2 <- cmdscale(D, k = 2)
B2 <- as.data.frame(cbind(d, m2))

ggplot(B2, aes(x = `1`, y = `2`, shape = nationality, colour = `Prevotella melaninogenica et rel.`)) +
  geom_point(size = 4) +
  scico::scale_colour_scico(palette = "vik") +
  theme_minimal()

```





```{r}

h <- data.frame(d$nationality, A)

m3 <- MASS::lda(d.nationality ~ ., h, method = "mle")
m4 <- e1071::svm(d.nationality ~., h)


h$LD1 <- predict(m3, h)$x
h$lda_class <- predict(m3, h)$class
h$svm <- predict(m4, h)
h$bmi_group <- d$bmi_group

xtabs(~lda_class + d.nationality, h)
xtabs(~svm + d.nationality, h)


ggplot(h, aes(y = LD1, x = d.nationality, colour = bmi_group)) +
  ggbeeswarm::geom_quasirandom() +
  theme_minimal()

```


